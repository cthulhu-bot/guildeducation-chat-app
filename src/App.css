.App-messages {
  text-align: center;
}

.App-messageBox {
  margin-top: 1rem;
  width: 320px;
  letter-spacing: 0.1em !important;
  text-indent: 3px;
  height: 50px;
  border-radius: 0px !important;
  margin-bottom: 100px;
}

.App-messageButton {
  margin-left: 10px;
}