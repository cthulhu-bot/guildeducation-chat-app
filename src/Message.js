import React from "react";
import { createStyledComponent } from "mineral-ui/styles";
import Flex, { FlexItem } from "mineral-ui/Flex";

const LeftMessage = createStyledComponent(FlexItem, {
  background: "#0b93f6",
  borderRadius: "25px",
  color: "white",
  lineHeight: "24px",
  marginBottom: "12px",
  marginRight: "300px",
  marginTop: "0.5rem",
  maxWidth: "255px",
  padding: "10px 20px",
  textAlign: "center",
  width: "50%",
  wordWrap: "break-word"
});

const RightMessage = createStyledComponent(FlexItem, {
  background: "#e5e5ea",
  borderRadius: "25px",
  color: "black",
  lineHeight: "24px",
  marginBottom: "12px",
  marginLeft: "300px",
  marginTop: "0.5rem",
  maxWidth: "255px",
  padding: "10px 20px",
  textAlign: "center",
  width: "50%",
  wordWrap: "break-word"
});

const Message = ({ message }) => {
  return (
    <Flex justifyContent={"center"}>
      {message.name === "Jane Doe" ? (
        <LeftMessage
          data-testid={`message-${message.key}`}
          key={message.key}
        >{`${message.text}`}</LeftMessage>
      ) : (
        <RightMessage
          data-testid={`message-${message.key}`}
          key={message.key}
        >{`${message.text}`}</RightMessage>
      )}
    </Flex>
  );
};

export default Message;
