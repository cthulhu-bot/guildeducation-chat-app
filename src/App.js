import React, { Component } from "react";
import firebase from "./firebase";
import { ThemeProvider } from "mineral-ui/themes";
import Message from "./Message";
import "./App.css";

/**
 * Entering messages into the message box will be saved off and
 * redistributed to any other open tabs via the firebase subscription.
 * Entering a text message will randomly assign a hard coded user.
 */
class App extends Component {
  constructor() {
    super();
    this.state = {
      currentVal: "",
      messages: []
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   * componentDidMount
   * Query for messages when the component first mounts
   */
  componentDidMount() {
    this.loadMessages();
  }

  handleChange(event) {
    this.setState({ currentVal: event.target.value });
  }

  handleSubmit(event) {
    event.preventDefault();
    if (!event.target[0].value.trim()) {
      return;
    }
    this.props.saveMessage
      ? this.props.saveMessage(event.target[0].value)
      : this.saveMessage(event.target[0].value);
    this.setState({ currentVal: "" });
  }

  /**
   * loadMessages
   * Registers the 2 firebase listeners: child_added and child_changed which
   * call updateMessages when something is updated in firebase which in turn
   * updates component state and kicks off the render cycle
   */
  loadMessages() {
    const updateMessages = snapshot => {
      const data = snapshot.val();
      const newMessage = {
        key: snapshot.key,
        name: data.name,
        text: data.text
      };
      const messages = this.state.messages;
      messages.push(newMessage);

      this.setState({
        messages: messages
      });
    };

    // listen for changes in firebase and updateState accordingly
    firebase
      .database()
      .ref("/messages/")
      .limitToLast(8)
      .on("child_added", updateMessages);
    firebase
      .database()
      .ref("/messages/")
      .limitToLast(8)
      .on("child_changed", updateMessages);
  }

  /**
   * saveMessage
   * Pushes messages from the textInput out to firebase
   * @param messageText
   */
  saveMessage(messageText) {
    const name = Math.floor(Math.random() * 2) === 0 ? "John Doe" : "Jane Doe";
    return firebase
      .database()
      .ref("/messages/")
      .push({
        name: name,
        text: messageText
      })
      .catch(function(error) {
        console.error("Error writing new message to firebase", error);
      });
  }

  render() {
    return (
      <ThemeProvider>
        <div className="App-messages">
          {this.state.messages.map(message => (
            <Message message={message} />
          ))}
          <form onSubmit={this.handleSubmit}>
            <input
              value={this.state.currentVal}
              onChange={this.handleChange}
              data-testid="input"
              className="App-messageBox"
            />
            <button
              className="App-messageButton"
              data-testid="inputButton"
              type="submit"
            >
              ^
            </button>
          </form>
        </div>
      </ThemeProvider>
    );
  }
}

export default App;
