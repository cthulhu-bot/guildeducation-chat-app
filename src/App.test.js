import React from "react";
import App from "./App";
import renderer from "react-test-renderer";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Message from "./Message";

Enzyme.configure({ adapter: new Adapter() });

describe("Messanger app", () => {
  it("should render the form inputs with 0 messages and without throwing an error", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("input[data-testid='input']").length).toEqual(1);
    expect(wrapper.find("button[data-testid='inputButton']").length).toEqual(1);
    expect(wrapper.state()).toEqual({ currentVal: "", messages: [] });
    expect(wrapper.find(Message).length).toEqual(0);
  });

  it("should update its component state to respond to change events", () => {
    const saveMessage = jest.fn();
    const wrapper = mount(<App saveMessage={saveMessage} />);
    wrapper
      .find("input[data-testid='input']")
      .simulate("change", { target: { value: "Hello World!" } });
    expect(wrapper.state().currentVal).toEqual("Hello World!");
  });

  it("should match its own snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
