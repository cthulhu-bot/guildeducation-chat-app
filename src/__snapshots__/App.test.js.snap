// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messanger app should match its own snapshot 1`] = `
<div
  className="App-messages"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="App-messageBox"
      data-testid="input"
      onChange={[Function]}
      value=""
    />
    <button
      className="App-messageButton"
      data-testid="inputButton"
      type="submit"
    >
      ^
    </button>
  </form>
</div>
`;
